crates/plugins/src/badger/store.rs

```diff
@@ -0,0 +1,103 @@
+use std::path::PathBuf;
+
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+
+const DEFAULT_STORE_DIR: &str = "spin";
+const DEFAULT_STORE_FILE: &str = "plugins-badger.json";
+
+pub struct BadgerRecordManager {
+    db_path: PathBuf,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct BadgerRecord {
+    name: String,
+    badgered_from: semver::Version,
+    badgered_to: Vec<semver::Version>,
+    when: chrono::DateTime<chrono::Utc>,
+}
+
+pub enum PreviousBadger {
+    Fresh,
+    FromCurrent {
+        to: Vec<semver::Version>,
+        when: chrono::DateTime<chrono::Utc>,
+    },
+}
+
+impl PreviousBadger {
+    fn includes(&self, version: &semver::Version) -> bool {
+        match self {
+            Self::Fresh => false,
+            Self::FromCurrent { to, .. } => to.contains(version),
+        }
+    }
+
+    pub fn includes_any(&self, version: &[&semver::Version]) -> bool {
+        version.iter().any(|version| self.includes(version))
+    }
+}
+
+impl BadgerRecordManager {
+    pub fn default() -> anyhow::Result<Self> {
+        let base_dir = dirs::cache_dir()
+            .or_else(|| dirs::home_dir().map(|p| p.join(".spin")))
+            .ok_or_else(|| anyhow!("Unable to get local data directory or home directory"))?;
+        let db_path = base_dir.join(DEFAULT_STORE_DIR).join(DEFAULT_STORE_FILE);
+        Ok(Self { db_path })
+    }
+
+    fn load(&self) -> Vec<BadgerRecord> {
+        match std::fs::read(&self.db_path) {
+            Ok(v) => serde_json::from_slice(&v).unwrap_or_default(),
+            Err(_) => vec![],
```

Comment:
I did ponder that, but couldn't think of anything useful to do at this point.  If the shape changes incompatibly, our choices are:

1. Adapt it. We can't do that until the change happens, that would be a matter for this function anyway.
2. Clear the decks and go "we can't understand this ancient file, let's assume we need to re-check everything," which is what this would do if left unchanged.

Could you expand on what additional error handling you're thinking of?

Reply:
I suppose if it is expected to be immediately overwritten there isn't much to do.

Reply:
Possibly worth a comment? Silently ignoring errors usually activates my bug-dar.